src/docs/rts-new.md

```diff
@@ -0,0 +1,170 @@
+# Interaction with timestamps
+
+The main thing RTS is concerned with are the durable and stable
+timestamps. The high-level details are covered above, what follows is a more
+detailed view of how we use these timestamps.
+
+We read `txn_global` for pinned and stable timestamps at the start of RTS,
+and in diagnostic builds we read them again at the end to assert they didn't
+change underneath us (indicating the system was not quiescent, and therefore
+a bug).
+
+Once RTS has finished, we roll back the global durable timestamp to the
+stable timestamp, since if there are no unstable updates, `txn_global`'s
+durable timestamp can by definition be moved backwards to the stable
+timestamp. If we weren't to do this, it wouldn't automatically be a problem,
+but we'd likely end up unnecessarily scanning tables the next time RTS is
+run.
+
+Timestamps of updates in the history store can have some surprising
+properties:
+
+* The start timestamp can be equal to the stop timestamp if the original
+  update's commit timestamp is in order.
+* During rollback of a prepared transaction, a history store update with a
+  stop timestamp may not get removed, leading to a duplocate record. RTS
+  ignores these.
+* If we've had to fix any timestamps as part of RTS, a newly inserted update
+  with an older timestamp may have a durable timestamp smaller than the
+  current stop durable timestamp.
+* When a checkpoint writes a committed prepared update, it's possible for
+  further updates on that key to include history store changes before the
+  transaction alters the history store update to have a proper stop
+  timestamp. Thus, we can have an update in the history store with a maximum
+  stop timestamp in the middle of other updates for the same key.
+  
+We also have some verification specifically related to the history
+store. While iterating over records, we check that the start time of a newly
+encountered record is greater than or equal to the previous record. We also
+validate that the timestamps in the key and the cell are the same (where
+possible).
+
+Fast-truncate is worth mentioning briefly. If the page has a `page_del`
+structure, we can look at its `durable_timestamp`, much the same as we would
+a normal page.
+
+# Interaction with transaction IDs
+
+This is a little simpler than the timestamp rules - we use normal snapshot
+visibility in most of the places we look at timestamps (see the timestamps
+section). One exception is that we treat all data as visible when RTS is
+called from the API (as opposed to startup or shutdown). 
+
+We also clear the transaction ID of updates we look at during RTS, because
+the connection's write generation will be initialised after RTS and the
+updates in the cache would be problematic for other parts of the code if
+they had a "legitimate" transaction ID.
```

Comment:
@kommiharibabu it wasn't clear from the code why this is the case, do you know why?

Reply:
Thanks - I've added the example we discussed in ae692da.